go/plugins/postgresql/docstore.go

```diff
@@ -0,0 +1,157 @@
+}
+
+func (ds *docStore) query(ctx context.Context, ss *SimilaritySearch, embbeding []float32) (*ai.RetrieverResponse, error) {
+	res := &ai.RetrieverResponse{}
+
+	query := ss.buildQuery(embbeding)
+	rows, err := ds.engine.Pool.Query(ctx, query)
+	if err != nil {
+		return nil, err
+	}
+	defer rows.Close()
+	fieldDescriptions := rows.FieldDescriptions()
+	columnNames := make([]string, len(fieldDescriptions))
+
+	for i, fieldDescription := range fieldDescriptions {
+		columnNames[i] = fieldDescription.Name
+	}
+
+	for rows.Next() {
+		values := make([]interface{}, len(columnNames))
+		valuesPrt := make([]interface{}, len(columnNames))
+
+		for i := range columnNames {
+			valuesPrt[i] = &values[i]
+		}
+		if err := rows.Scan(valuesPrt...); err != nil {
+			return nil, fmt.Errorf("scan row failed: %v", err)
+		}
+
+		meta := make(map[string]any, ss.k)
+		var content []*ai.Part
+		for i, col := range columnNames {
+			if (len(ss.metadataColumns) > 0 && !slices.Contains(ss.metadataColumns, col)) &&
+				ss.contentColumn != col &&
+				ss.metadataJsonColumn != col {
+				continue
+			}
+
+			if ss.contentColumn == col {
+				content = append(content, ai.NewTextPart(values[i].(string)))
+			}
+
+			if ss.metadataJsonColumn == col {
+				content = append(content, ai.NewJSONPart(values[i].(string)))
+				continue
+			}
+
+			meta[col] = values[i]
+		}
+
+		doc := &ai.Document{
+			Metadata: meta,
+			Content:  content,
+		}
+
+		res.Documents = append(res.Documents, doc)
+	}
+
+	return res, nil
+}
+
+type RetrieverOptions struct {
+	SimilaritySearch *SimilaritySearch
+}
+
+func (ds *docStore) Retrieve(ctx context.Context, req *ai.RetrieverRequest) (*ai.RetrieverResponse, error) {
+	if req.Options == nil {
+		ss := NewSimilaritySearch()
+		req.Options = &RetrieverOptions{SimilaritySearch: ss}
+	}
+
+	ropt, ok := req.Options.(*RetrieverOptions)
+	if !ok {
+		return nil, fmt.Errorf("postgres.Retrieve options have type %T, want %T", req.Options, &RetrieverOptions{})
+	}
+
+	if ropt.SimilaritySearch == nil {
+		ss := NewSimilaritySearch()
+		ropt.SimilaritySearch = ss
+	}
+
+	ereq := &ai.EmbedRequest{
+		Documents: []*ai.Document{req.Query},
+		Options:   ds.embedderOptions,
+	}
+	eres, err := ds.embedder.Embed(ctx, ereq)
+	if err != nil {
+		return nil, fmt.Errorf("postgres.Retrieve retrieve embedding failed: %v", err)
+	}
+	res, err := ds.query(ctx, ropt.SimilaritySearch, eres.Embeddings[0].Embedding)
+	if err != nil {
+		return nil, fmt.Errorf("googlecloudsql.postgres.Retrieve failed to execute the query: %v", err)
+	}
+	return res, nil
+}
+
+// Index implements the genkit Retriever.Index method.
+func (ds *docStore) Index(ctx context.Context, req *ai.IndexerRequest) error {
+	if len(req.Documents) == 0 {
+		return nil
+	}
+	//TODO: implement method
+	return nil
+}
+
+// newDocStore instantiate a docStore
+func newDocStore(ctx context.Context, p *Postgres, cfg Config) (*docStore, error) {
+	p.mu.Lock()
+	defer p.mu.Unlock()
+	if !p.initted {
```

Comment:
This method should make sure  the table columns exist and set all fields of the Docstore. Here is an example in Python:
https://github.com/googleapis/langchain-google-alloydb-pg-python/blob/9733c439c14bccceb1d7a6538cc21bee1c1d87af/src/langchain_google_alloydb_pg/async_vectorstore.py#L182-L231